### PR TITLE
app-emulation/libguestfs: libxcrypt, py3 updates

### PR DESCRIPTION
**Pull request by hlein**
Current version won't build with libxcrypt 4.x, which is currently
being stabilized.  Workaround taken from b.g.o; also bump supported
Python versions and EAPI version.

Signed-off-by: Hank Leininger <hlein@korelogic.com>
Closes: https://bugs.gentoo.org/703118
Bug: https://bugs.gentoo.org/734634
Package-Manager: Portage-3.0.0, Repoman-2.3.23

### CODE REVIEW

**Comment by thesamesam** [comment]
Fancy rebasing this?

**Comment by hlein** [comment]
> Fancy rebasing this if it's still relevant?

Hm. Yeah, sure. The current ebuild in tree has already been converted to `EAPI=7`, so that change is no longer needed, but still won't build with current libxcrypt, so that part is worth keeping.

However, the current in-tree version also had Python support completely ripped out in https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=34e0d817f35aad31d992a35f436603e3e8eed3bb , and I don't know why, maybe because it was only 3.6? I've had no trouble with 3.7, which is why I created this PR months ago. Or 3.8 for that matter. So I could re-add python, with the bump, if that's the only reason it was removed.

The in-tree version is also missing the `COMMON_DEPEND` change from `dev-ml/ocaml-gettext` to `dev-ml/ocaml-gettext-stub`; I could include that unless it's already in another pending PR elsewhere.

**Comment by thesamesam** [comment]
I'm about to merge the OCaml changes from [here](https://github.com/gentoo/gentoo/pull/19095) which includes that. I think we can restore Python if it's useful? We yanked it because of the lack of 3.7+ support, that's all.

The libxcrypt part is definitely worth keeping. Once I merge yours, I'll do a general tidy up too as there's a lot of open bugs for this. Feel free to fix any that take your fancy, especially as this is now maintainer-needed :(

**Comment by hlein** [comment]
OK @thesamesam please take a look - restored python and bumped compatible versions, and applied the libxcrypt fix. Thanks!

**Comment by ajakk** [comment]
Instead of adding a patch for the libcrypt bug, why not `append-ldflags '-L/usr/$(get_libdir)/xcrypt'`? One ebuild line will hopefully be easier to maintain than a patch.

**Comment by thesamesam** [comment]
(Note that for the patch, we need to use @LIBDIR@ or similar and sed it later. append-ldflags sounds easier.)

**Comment by hlein** [comment]
> Instead of adding a patch for the libcrypt bug, why not `append-ldflags '-L/usr/$(get_libdir)/xcrypt'`? One ebuild line will hopefully be easier to maintain than a patch.

tl;dr: because wonky build system.

That sounded reasonable, so I tried it:

```diff --git a/app-emulation/libguestfs/libguestfs-1.38.6-r103.ebuild b/app-emulation/libguestfs/libguestfs-1.38.6-r1>
index 04a96b99..32805fa4 100644
--- a/app-emulation/libguestfs/libguestfs-1.38.6-r103.ebuild
+++ b/app-emulation/libguestfs/libguestfs-1.38.6-r103.ebuild
@@ -138,6 +138,7 @@ src_configure() {
        # configured kernel.
        export vmchannel_test=no
 
+       append-ldflags '-L/usr/$(get_libdir)/xcrypt'
        econf \
                --with-bashcompletiondir="$(get_bashcompdir)" \
                $(use_with libvirt) \
```

...assuming that is what you meant. And removed `0003-Fix-libxcrypt.patch`, of course.

Building gets quite some ways, but then fails with:

```../ocaml-link.sh -cclib '-lvisit -lstructs -lutils -ltinfo -L/usr/lib64/xcrypt -lcrypt -lxml2  -lgnu' -- ocamlfind opt -g -annot -safe-string -warn-error CDEFLMPSUVYZX+52-3 -ccopt '-march=znver2 -freorder-blocks-and-partition -O2 -pipe -msse4.1 -msse4.2 -msse4 -mavx -fno-strict-overflow -Wno-strict-overflow' -package str,unix -I ../common/utils/.libs -I ../common/structs/.libs -I ../lib/.libs -I ../gnulib/lib/.libs -I ../ocaml -I ../common/visit/.libs -I ../common/mlstdutils -I ../common/mlutils -I ../common/mlgettext -I ../common/mlpcre -I ../common/mlvisit -I ../common/mltools -I ../customize -package gettext-stub mlstdutils.cmxa mlguestfs.cmxa mlgettext.cmxa mlpcre.cmxa mlcutils.cmxa mltools.cmxa mlvisit.cmxa customize.cmxa  utils.cmx sysprep_operation.cmx sysprep_operation_abrt_data.cmx sysprep_operation_backup_files.cmx sysprep_operation_bash_history.cmx sysprep_operation_blkid_tab.cmx sysprep_operation_ca_certificates.cmx sysprep_operation_crash_data.cmx sysprep_operation_cron_spool.cmx sysprep_operation_customize.cmx sysprep_operation_dhcp_client_state.cmx sysprep_operation_dhcp_server_state.cmx sysprep_operation_dovecot_data.cmx sysprep_operation_flag_reconfiguration.cmx sysprep_operation_firewall_rules.cmx sysprep_operation_fs_uuids.cmx sysprep_operation_kerberos_data.cmx sysprep_operation_lvm_uuids.cmx sysprep_operation_logfiles.cmx sysprep_operation_machine_id.cmx sysprep_operation_mail_spool.cmx sysprep_operation_net_hostname.cmx sysprep_operation_net_hwaddr.cmx sysprep_operation_pacct_log.cmx sysprep_operation_package_manager_cache.cmx sysprep_operation_pam_data.cmx sysprep_operation_passwd_backups.cmx sysprep_operation_puppet_data_log.cmx sysprep_operation_rh_subscription_manager.cmx sysprep_operation_rhn_systemid.cmx sysprep_operation_rpm_db.cmx sysprep_operation_samba_db_log.cmx sysprep_operation_script.cmx sysprep_operation_smolt_uuid.cmx sysprep_operation_ssh_hostkeys.cmx sysprep_operation_ssh_userdir.cmx sysprep_operation_sssd_db_log.cmx sysprep_operation_tmp_files.cmx sysprep_operation_udev_persistent_net.cmx sysprep_operation_user_account.cmx sysprep_operation_utmp.cmx sysprep_operation_yum_uuid.cmx main.cmx -o virt-sysprep virt_sysprep-dummy.o  
rm -f sysprep-extra-options.pod sysprep-extra-options.pod-t
rm -f sysprep-operations.pod sysprep-operations.pod-t
../run virt-sysprep --dump-pod-options > sysprep-extra-options.pod-t
../run virt-sysprep --dump-pod > sysprep-operations.pod-t
virt-sysprep: error while loading shared libraries: libcrypt.so.2: cannot open shared object file: No such file or directory
make[2]: *** [Makefile:2738: sysprep-extra-options.pod] Error 127
make[2]: *** Waiting for unfinished jobs....
virt-sysprep: error while loading shared libraries: libcrypt.so.2: cannot open shared object file: No such file or directory
make[2]: *** [Makefile:2743: sysprep-operations.pod] Error 127
make[2]: Leaving directory '/var/tmp/portage/app-emulation/libguestfs-1.38.6-r103/work/libguestfs-1.38.6/sysprep'
make[1]: *** [Makefile:2225: all-recursive] Error 1
make[1]: Leaving directory '/var/tmp/portage/app-emulation/libguestfs-1.38.6-r103/work/libguestfs-1.38.6'
make: *** [Makefile:2129: all] Error 2
```

`run` is a shell script at the toplevel, built from `run.in` I think; looks like it handles/respects `LD_LIBRARY_PATH` but does not pick up `append-ldflags` changes. Maybe we could run around patching run.in and whatever else needs it, but then it's no longer a clean alternative.

Another b.g.o commenter suggested adding `/usr/lib64/xcrypt/` (or I suppose, `/usr/$(get_libdir)/xcrypt`) to an `/etc/ld.so.conf.d/` file, which allows building with no other changes. At first glance that looked like a good idea to me, except 1) doing so in the libguestfs ebuild is too late, because it needs to be in place before we build; 2) if it was really a good idea, why didn't the libxcrypt package do it already? So probably there is a reason why not.

**Comment by ajakk** [comment]
The flag-o-matic eclass provides append-ldflags, I think you forgot to inherit it.

**Comment by hlein** [comment]
> The flag-o-matic eclass provides append-ldflags, I think you forgot to inherit it.

You are right, I sure did!

...After fixing that, and changing the append-ldflags call to double-quote instead of single quote, it works. Will update the PR in a minute.

**Comment by GuillaumeSeren** [comment]
Hey @hlein,
I've run into issue 703118 and found this PR, I tested it on my system,
it works fine without workaround, thank you !

**Comment by thesamesam** [comment]
That seems really weird. I wonder if LDFLAGS is not being fully respected...?

**Comment by hlein** [comment]
> That seems really weird. I wonder if LDFLAGS is not being fully respected...?

@thesamesam I believe it works as of the updated PR from Jan 24, just waiting for review & merge.

**Comment by ajakk** [comment]
What about the 1.36 branch? I'm not sure if there's any reason to keep it, so if it won't get these fixes it might be a good idea to clean it up?

**Comment by hlein** [comment]
> What about the 1.36 branch? I'm not sure if there's any reason to keep it, so if it won't get these fixes it might be a good idea to clean it up?

I would love to propose removing 1.36, which hasn't gotten an update from upstream in over 2 years, and maybe submit a PR adding an ebuild for 1.40 (the current stable branch).

But first I would just like to get this landed, after 6 months of trying.

**Comment by ajakk** [comment]
That's understandable, but I'm not sure it's sane to close a bug when there are still affected versions in tree. I imagine we can get by with a revbump of each branch with these fixes, and then a cleanup of affected versions to close the bug. We have a reason to cleanup anyway due to the dependency on unslotted lua.

**Comment by hlein** [comment]
> That's understandable, but I'm not sure it's sane to close a bug when there are still affected versions in tree. I imagine we can get by with a revbump of each branch with these fixes, and then a cleanup of affected versions to close the bug. We have a reason to cleanup anyway due to the dependency on unslotted lua.

Fair enough. OK, in fact dependencies for libguestfs-1.36 were removed in d0c2104377b00b4b6b029ba2e6390a00c58057b2, so I've just dropped them; PR rebased and updated.

**Comment by thesamesam** [comment]
Thank you!
